apps/api/src/app/inbox/usecases/update-notification/update-notification.usecase.ts

```diff
@@ -0,0 +1,117 @@
+} from '@novu/application-generic';
+import { MessageEntity, MessageRepository } from '@novu/dal';
+import { ButtonTypeEnum, MessageActionStatusEnum } from '@novu/shared';
+
+import { ApiException } from '../../../shared/exceptions/api.exception';
+import { GetSubscriber } from '../../../subscribers/usecases/get-subscriber';
+import { AnalyticsEventsEnum } from '../../utils';
+import { mapToDto } from '../../utils/notification-mapper';
+import { InboxNotification } from '../../utils/types';
+import type { UpdateNotificationCommand } from './update-notification.command';
+
+@Injectable()
+export class UpdateNotification {
+  constructor(
+    private invalidateCache: InvalidateCacheService,
+    private getSubscriber: GetSubscriber,
+    private analyticsService: AnalyticsService,
+    private messageRepository: MessageRepository
+  ) {}
+
+  async execute(command: UpdateNotificationCommand): Promise<InboxNotification> {
+    const subscriber = await this.getSubscriber.execute({
+      environmentId: command.environmentId,
+      organizationId: command.organizationId,
+      subscriberId: command.subscriberId,
+    });
+    if (!subscriber) {
+      throw new ApiException(`Subscriber with id: ${command.subscriberId} is not found.`);
+    }
+
+    const message = await this.messageRepository.findOne({
+      _environmentId: command.environmentId,
+      _subscriberId: subscriber._id,
+      _id: command.notificationId,
+    });
+    if (!message) {
+      throw new NotFoundException(`Notification with id: ${command.notificationId} is not found.`);
+    }
+
+    const isUpdatingPrimaryCta = command.primaryActionCompleted !== undefined;
+    const isUpdatingSecondaryCta = command.secondaryActionCompleted !== undefined;
+    const primaryCta = message.cta.action?.buttons?.find((button) => button.type === ButtonTypeEnum.PRIMARY);
+    const secondaryCta = message.cta.action?.buttons?.find((button) => button.type === ButtonTypeEnum.SECONDARY);
+    if ((isUpdatingPrimaryCta && !primaryCta) || (isUpdatingSecondaryCta && !secondaryCta)) {
+      throw new ApiException(
+        `Could not perform action on the ${isUpdatingPrimaryCta && !primaryCta ? 'primary' : 'secondary'} button.`
+      );
+    }
+
+    const updatePayload: Partial<MessageEntity> = {
+      ...(command.read !== undefined && { read: command.read }),
+      ...(command.archived !== undefined && { archived: command.archived }),
+    };
+
+    if (isUpdatingPrimaryCta) {
+      updatePayload['cta.action.result.type'] = ButtonTypeEnum.PRIMARY;
+      updatePayload['cta.action.status'] = command.primaryActionCompleted
+        ? MessageActionStatusEnum.DONE
+        : MessageActionStatusEnum.PENDING;
+    }
+
+    if (isUpdatingSecondaryCta) {
+      updatePayload['cta.action.result.type'] = ButtonTypeEnum.SECONDARY;
+      updatePayload['cta.action.status'] = command.secondaryActionCompleted
+        ? MessageActionStatusEnum.DONE
+        : MessageActionStatusEnum.PENDING;
+    }
```

Comment:
The current implementation for the primary/secondary buttons is weird and requires refactoring. 
But in short, we only allow one button to have its status changed to "done".